### ISSUE TITLE
`manual_retain`: incorrect suggestions for some closure parameter patterns  

### ISSUE THREAD

**Issue by yuxqiu** [rust-lang/rust-clippy#12081]
### Summary

The `manual_retain` lint fails to be applied when a closure given to the `filter` method uses a `&` in the closure argument pattern. For example, `filter(|&x| x == 0)`.

### Reproducer

I tried this code:

```rust
fn main() {
    let mut v = vec![1, 2, 3];
    v = v.into_iter().filter(|&x| x == 0).collect();
}
```

I expected to see that the lint gives the following suggestion:

```rust
fn main() {
    let mut v = vec![1, 2, 3];
    v.retain(|&x| x == 0);
}
```

Instead, the suggestion failed to apply as the lint incorrectly suggests:

```rust
fn main() {
    let mut v = vec![1, 2, 3];
    v.retain(|x| x == 0);
}
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by yuxqiu** [comment]
After some investigation, I found that the following code may also trigger incorrect suggestions:

```rust
fn main() {
    let mut v = vec![1, 2, 3];
    v = v.iter().filter(|x| **x == 0).copied().collect();
}
```

```rust
fn main() {
    let mut s = String::from("abc");
    s = s.chars().filter(|c| *c == 'a').to_owned().collect();
}
```

The main reason for this is that the closure is given directly to the `retain` function (`v.retain(|x| **x == 0)`). This works for `into_iter().filter(<f>)`, since both the `retain` method and the `filter` method here use a predicate bounded by `FnMut(&Self::Item) -> bool`.

However, this is not the case for `iter().filter(<f>)` ([iter-filter-doc](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter)) and `chars().filter(<f>)` ([chars-filter-doc](https://doc.rust-lang.org/std/str/struct.Chars.html#method.filter) and [string-retain-doc](https://doc.rust-lang.org/std/string/struct.String.html#method.retain)). For these two cases, we need to "strip" a layer of references when processing the lambda used by the `filter` method.
